hibernate-core/src/main/java/org/hibernate/jpa/boot/internal/EntityManagerFactoryBuilderImpl.java

```diff
@@ -199,74 +202,81 @@ private EntityManagerFactoryBuilderImpl(
-				metadataSources,
-				metamodelBuilder.getBootstrapContext()
-		);
+			this.managedResources = MetadataBuildingProcess.prepare(
+					metadataSources,
+					metamodelBuilder.getBootstrapContext()
+			);
 
-		applyMetadataBuilderContributor();
+			applyMetadataBuilderContributor();
 
 
-		withValidatorFactory( configurationValues.get( org.hibernate.cfg.AvailableSettings.JPA_VALIDATION_FACTORY ) );
+			withValidatorFactory( configurationValues.get( org.hibernate.cfg.AvailableSettings.JPA_VALIDATION_FACTORY ) );
 
-		// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
-		// push back class transformation to the environment; for the time being this only has any effect in EE
-		// container situations, calling back into PersistenceUnitInfo#addClassTransformer
+			// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+			// push back class transformation to the environment; for the time being this only has any effect in EE
+			// container situations, calling back into PersistenceUnitInfo#addClassTransformer
 
-		final boolean dirtyTrackingEnabled = readBooleanConfigurationValue( AvailableSettings.ENHANCER_ENABLE_DIRTY_TRACKING );
-		final boolean lazyInitializationEnabled = readBooleanConfigurationValue( AvailableSettings.ENHANCER_ENABLE_LAZY_INITIALIZATION );
-		final boolean associationManagementEnabled = readBooleanConfigurationValue( AvailableSettings.ENHANCER_ENABLE_ASSOCIATION_MANAGEMENT );
+			final boolean dirtyTrackingEnabled = readBooleanConfigurationValue( AvailableSettings.ENHANCER_ENABLE_DIRTY_TRACKING );
+			final boolean lazyInitializationEnabled = readBooleanConfigurationValue( AvailableSettings.ENHANCER_ENABLE_LAZY_INITIALIZATION );
+			final boolean associationManagementEnabled = readBooleanConfigurationValue( AvailableSettings.ENHANCER_ENABLE_ASSOCIATION_MANAGEMENT );
 
-		if ( dirtyTrackingEnabled || lazyInitializationEnabled || associationManagementEnabled ) {
-			EnhancementContext enhancementContext = getEnhancementContext(
-					dirtyTrackingEnabled,
-					lazyInitializationEnabled,
-					associationManagementEnabled
-			);
+			if ( dirtyTrackingEnabled || lazyInitializationEnabled || associationManagementEnabled ) {
+				EnhancementContext enhancementContext = getEnhancementContext(
+						dirtyTrackingEnabled,
+						lazyInitializationEnabled,
+						associationManagementEnabled
+				);
 
-			persistenceUnit.pushClassTransformer( enhancementContext );
-		}
+				persistenceUnit.pushClassTransformer( enhancementContext );
+			}
 
-		// for the time being we want to revoke access to the temp ClassLoader if one was passed
-		metamodelBuilder.applyTempClassLoader( null );
+			// for the time being we want to revoke access to the temp ClassLoader if one was passed
+			metamodelBuilder.applyTempClassLoader( null );
+		}
+		catch (Throwable t) {
+			bsr.close();
+			cleanup();
```

Comment:
Here too, I think a try/catch + addSuppressed would be better; especially on 5.3 where we don't want backwards-incompatible changes.